Fix response schema for delete batch/v1/jobs

JoeryH opened the pull request:
DELETE batch/v1/jobs actually returns a Job resource,
not the Status the swagger docs currently have.

This causes the clients to be generated incorrectly and
the reference docs to be incorrect.

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change

/kind bug

> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

This PR updates the response schema for DELETE batch/v1/jobs in swagger.json.
We need this because swagger.json is currently incorrect (for at least v1.16, v1.17 and latest) and says a Status is returned instead of Job. This causes clients to be generated incorrectly and crashing on deserialisation.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
API reference for DELETE batch/v1/jobs now correctly states a Job is returned.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

JoeryH commented:
/assign @thockin

JoeryH commented:
/remove-lifecycle stale

thockin commented:
@janetkuo Can you help me find a home for this?  I got assigned but I have no real context :)

janetkuo reviewed (changes requested):
This file is generated and shouldn't be modified manually, ref https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md

JoeryH committed:
Fix response schema for delete batch/v1/jobs

DELETE batch/v1/jobs actually returns a Job resource,
not the generic Status object.

We can use ReturnDeletedObject: true in the storage
configuration to fix this.

Without this fix clients that are generated will be
incorrectly unmarshalling the json and, at least for
the Java client, throw an exception.

JoeryH committed:
Update generated swagger API

JoeryH commented:
Hi @janetkuo , thank you for your message. I found a related issue that made me find the required change so that the swagger doc is generated correctly for delete job. https://github.com/kubernetes/kubernetes/issues/59501

JoeryH commented:
Not sure what labels there should be, but just removing the bulk of them that were added.

/assign @janetkuo

/remove-area apiserver
/remove-area cloudprovider
/remove-area code-generation
/remove-area conformance
/remove-area dependency
/remove-area ipvs
/remove-area kubeadm
/remove-area kubectl
/remove-area kubelet
/remove-area provider/gcp
/remove-area release-eng
/remove-kind api-change
/remove-sig api-machinery
/remove-sig apps
/remove-sig auth
/remove-sig autoscaling
/remove-sig cli
/remove-sig cloud-provider
/remove-sig cluster-lifecycle
/remove-sig contributor-experience
/remove-sig instrumentation
/remove-sig network
/remove-sig node
/remove-sig release
/remove-sig scalability
/remove-sig scheduling
/remove-sig storage
/remove-sig testing

JoeryH commented:
/remove-lifecycle stale

liggitt commented on pkg/registry/batch/job/storage/storage.go:
```diff
@@ -68,6 +68,7 @@ func NewREST(optsGetter generic.RESTOptionsGetter) (*REST, *StatusREST, error) {
 		UpdateStrategy:      job.Strategy,
 		DeleteStrategy:      job.Strategy,
 		ResetFieldsStrategy: job.Strategy,
+		ReturnDeletedObject: true,
```
this actually changes the API behavior of deleting a single job... I don't think we want to do that

liggitt replied:
hmm... deleting a single job sometimes returning status, as advertised:
```
kubectl delete jobs --all --v=9
I0917 12:42:00.787591    6800 request.go:1181] Request Body: {"propagationPolicy":"Background"}
I0917 12:42:00.787648    6800 round_trippers.go:435] curl -v -XDELETE  -H "User-Agent: kubectl/v1.23.0 (darwin/amd64) kubernetes/5fb1442" -H "Accept: application/json" -H "Content-Type: application/json" 'https://localhost:6443/apis/batch/v1/namespaces/default/jobs/pi'
I0917 12:42:00.791982    6800 round_trippers.go:454] DELETE https://localhost:6443/apis/batch/v1/namespaces/default/jobs/pi 200 OK in 4 milliseconds
I0917 12:42:00.792008    6800 round_trippers.go:460] Response Headers:
I0917 12:42:00.792017    6800 round_trippers.go:463]     Content-Length: 168
I0917 12:42:00.792025    6800 round_trippers.go:463]     Date: Fri, 17 Sep 2021 16:42:00 GMT
I0917 12:42:00.792031    6800 round_trippers.go:463]     Audit-Id: 6182eca4-fab7-40b0-96ff-406c439e6539
I0917 12:42:00.792037    6800 round_trippers.go:463]     Cache-Control: no-cache, private
I0917 12:42:00.792043    6800 round_trippers.go:463]     Content-Type: application/json
I0917 12:42:00.792049    6800 round_trippers.go:463]     X-Kubernetes-Pf-Flowschema-Uid: f882718c-ee58-4367-baee-cefdd030ae39
I0917 12:42:00.792055    6800 round_trippers.go:463]     X-Kubernetes-Pf-Prioritylevel-Uid: 5d04e46f-af20-40f0-a8f6-bcd5093a5d39
I0917 12:42:00.792072    6800 request.go:1181] Response Body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Success","details":{"name":"pi","group":"batch","kind":"jobs","uid":"09946e4e-6ba2-4e95-8852-8b6d1ddeabdd"}}
job.batch "pi" deleted
```

but can also return the full job:

```
kubectl delete --raw /apis/batch/v1/namespaces/default/jobs/pi --v=9
I0917 12:40:58.372314    5319 loader.go:372] Config loaded from file:  /Users/liggitt/go/src/k8s.io/kubernetes/_output/certs/admin.kubeconfig
I0917 12:40:58.373235    5319 round_trippers.go:435] curl -v -XDELETE  -H "Accept: application/json, */*" -H "User-Agent: kubectl/v1.23.0 (darwin/amd64) kubernetes/5fb1442" 'https://localhost:6443/apis/batch/v1/namespaces/default/jobs/pi'
I0917 12:40:58.383272    5319 round_trippers.go:454] DELETE https://localhost:6443/apis/batch/v1/namespaces/default/jobs/pi 200 OK in 10 milliseconds
I0917 12:40:58.383309    5319 round_trippers.go:460] Response Headers:
I0917 12:40:58.383323    5319 round_trippers.go:463]     X-Kubernetes-Pf-Prioritylevel-Uid: 5d04e46f-af20-40f0-a8f6-bcd5093a5d39
I0917 12:40:58.383332    5319 round_trippers.go:463]     Content-Length: 2009
I0917 12:40:58.383337    5319 round_trippers.go:463]     Date: Fri, 17 Sep 2021 16:40:58 GMT
I0917 12:40:58.383341    5319 round_trippers.go:463]     Audit-Id: 95d873ed-b5cb-4f8b-90ea-50d1f03ff23a
I0917 12:40:58.383362    5319 round_trippers.go:463]     Cache-Control: no-cache, private
I0917 12:40:58.383367    5319 round_trippers.go:463]     Content-Type: application/json
I0917 12:40:58.383372    5319 round_trippers.go:463]     Warning: 299 - "child pods are preserved by default when jobs are deleted; set propagationPolicy=Background to remove them or set propagationPolicy=Orphan to suppress this warning"
I0917 12:40:58.383377    5319 round_trippers.go:463]     X-Kubernetes-Pf-Flowschema-Uid: f882718c-ee58-4367-baee-cefdd030ae39
Warning: child pods are preserved by default when jobs are deleted; set propagationPolicy=Background to remove them or set propagationPolicy=Orphan to suppress this warning
{"kind":"Job","apiVersion":"batch/v1","metadata":{"name":"pi","namespace":"default","uid":"907a1e0e-f72a-4e42-9988-560925192ba3","resourceVersion":"436","generation":2,"creationTimestamp":"2021-09-17T16:40:50Z","deletionTimestamp":"2021-09-17T16:40:58Z","deletionGracePeriodSeconds":0,"labels":{"controller-uid":"907a1e0e-f72a-4e42-9988-560925192ba3","job-name":"pi"},"finalizers":["orphan"],"managedFields":[{"manager":"kube-controller-manager","operation":"Update","apiVersion":"batch/v1","time":"2021-09-17T16:40:50Z","fieldsType":"FieldsV1","fieldsV1":{"f:status":{"f:active":{},"f:startTime":{}}},"subresource":"status"},{"manager":"kubectl-create","operation":"Update","apiVersion":"batch/v1","time":"2021-09-17T16:40:50Z","fieldsType":"FieldsV1","fieldsV1":{"f:spec":{"f:backoffLimit":{},"f:completionMode":{},"f:completions":{},"f:parallelism":{},"f:suspend":{},"f:template":{"f:metadata":{"f:name":{}},"f:spec":{"f:containers":{"k:{\"name\":\"pi\"}":{".":{},"f:command":{},"f:image":{},"f:imagePullPolicy":{},"f:name":{},"f:resources":{},"f:terminationMessagePath":{},"f:terminationMessagePolicy":{}}},"f:dnsPolicy":{},"f:restartPolicy":{},"f:schedulerName":{},"f:securityContext":{},"f:terminationGracePeriodSeconds":{}}}}}}]},"spec":{"parallelism":1,"completions":1,"backoffLimit":6,"selector":{"matchLabels":{"controller-uid":"907a1e0e-f72a-4e42-9988-560925192ba3"}},"template":{"metadata":{"name":"pi","creationTimestamp":null,"labels":{"controller-uid":"907a1e0e-f72a-4e42-9988-560925192ba3","job-name":"pi"}},"spec":{"containers":[{"name":"pi","image":"perl","command":["perl","-Mbignum=bpi","-wle","print bpi(2000)"],"resources":{},"terminationMessagePath":"/dev/termination-log","terminationMessagePolicy":"File","imagePullPolicy":"Always"}],"restartPolicy":"Never","terminationGracePeriodSeconds":30,"dnsPolicy":"ClusterFirst","securityContext":{},"schedulerName":"default-scheduler"}},"completionMode":"NonIndexed","suspend":false},"status":{"startTime":"2021-09-17T16:40:50Z","active":1}}
```

liggitt replied:
It looks like the difference is whether or not propagationPolicy=Background is specified.

When it isn't, Status is returned. When it is, Job is returned.

JoeryH replied:
Ah I think I'm specifying propagationPolicy=Foregound.
Any reason to not always return a Job? Most clients crash right now because of type mismatch.

liggitt replied:
well... some clients generated to expect status and not specifying `propagationPolicy` are successfully working ... switching to always return a Job will break them...

JoeryH replied:
Okay thanks for looking into it, will close this then.

liggitt replied:
I think it's worth trying to improve the schema to indicate the object type is a possible return schema for the delete endpoints, even if we don't change the server behavior for existing types

liggitt commented:
There's a few things going on here:

1. Many (most?) types do not specify ReturnDeletedObject, not just Jobs
2. For those types, the generated schema specifies `Status` as the return type for deleting a single item and collection of items

Actual behavior of deleting a single item :
* Deleting a job with `propagationPolicy=Orphan`, `propagationPolicy=Foreground`, or  no `propagationPolicy` (which defaults to orphan) returns `Job`
* Deleting a job with `propagationPolicy=Background` returns `Status`
* Deleting a validationwebhook with `propagationPolicy=Orphan` or `propagationPolicy=Foreground` return `ValidatingWebhookConfiguration`
* Deleting a validationwebhook with `propagationPolicy=Background` or no `propagationPolicy` (which defaults to background) returns `Status`

Actual behavior of deleting a collection of items:
* Deleting a collection of jobs returns `JobList` in all cases
* Deleting a collection of validating webhooks returns `ValidatingWebhookConfigurationList` in all cases

So, for objects which do not set `ReturnDeletedObject: true`, the generated schema is incorrect in some cases for the single-item delete, and incorrect in all cases for the collection delete.

However, if we change to set `ReturnDeletedObject: true` for all objects, we will start returning objects in some cases where we previous returned Status, and will break clients successfully using those deletion paths that could not handle an object return.

Since all clients have to be able to handle Status in all cases (since it can be returned for forbidden errors, not found errors, etc), I think the safest thing to do is to make schema generation stop paying attention to `ReturnDeletedObject` and have it indicate the object can be a potential return value for single-item and collection deletion.

liggitt commented:
hoisted to https://github.com/kubernetes/kubernetes/issues/105104 to avoid losing the investigation and details
